[9.x] Add class support for console routing

chu121su12 opened the pull request:
Currently, `Artisan::command()` only supports closure as the second parameter. With this PR, it enables http route-style such as:
```php
Artisan::command('cli', [ClassName::class, '__invoke'])
```

~I'd like to add some test but I'm not sure how considering the `ClosureCommand` this class is mostly based on doesn't have test associated with it.~ Found the tests

Code for callable class binding is from https://stackoverflow.com/a/71010794

TBlindaruk committed:
[8.x] Added changelog

spawnia committed:
Test and annotate that Arr::pull() works with int keys (#42927)

This pull request corrects the type hint of the argument
`$key` in `Illuminate\Support\Arr::pull()` to allow
int keys. This avoids false-positives from IDEs and
static analysis tools.

In order to ensure it actually works as intended,
I added a test case.

josue-itg committed:
Fixed bug on forceCreate on a MorphMay relationship  not including morph type

josue-itg committed:
Added test for morph many not including type bug

josue-itg committed:
Style fixes on MorphMany.php

taylorotwell committed:
formatting

rodrigopedra committed:
[8.x] Handle cursor paginator when no items are found (#42963)

* handle cursor paginator when no items are found

* StyleCI suugested fix

StyleCIBot committed:
Apply fixes from StyleCI

BrandonSurowiec committed:
[Fix] Str::Mask() for consecutively repeating chars (#42295) (#42956)

- Add some additional test cases to verify behaviour for
  more than 2 repeating characters.
- Add some cases for testing the masking from the start of
  the string, before the start of the string and end of string.

(cherry picked from commit a91752882cbd1123d83847183441615f28928655)

Co-authored-by: Remy Bos <27890746+sjokkateer@users.noreply.github.com>

chu121su12 committed:
[9.x] Add support for `Artisan::command('cli', [ClassName::class, '__invoke'])`

chu121su12 committed:
remove static forward

taylorotwell commented:
If you are making a class anyways I would just define the signature in the class and register it like normal.

buismaarten committed:
explode(): Passing null to parameter #2 ($string) of type string is deprecated (#42980)

chu121su12 commented:
> If you are making a class anyways I would just define the signature in the class and register it like normal?

I've been doing that but one needs to register the class in the `Console\Kernel` instead. This adds api surface to allow callable class--in addition to closure--to be registered in the `routes/console.php`, similar to http routes.

chu121su12 committed:
add test

chu121su12 commented:
Added some test, but it seems the class binding failed in php 8.1.

michael-rubel committed:
[9.x] Fix BC break for `Log` feature tests (#42987)

* fix logger tests

* update tests for custom log levels

* remove unused import

* revert an old behavior for original log levels in tests

taylorotwell committed:
version

taylorotwell committed:
fix conflicts

jamescodium committed:
Fix forceCreate on MorphMany not returning newly created object (#42996)

TBlindaruk committed:
[8.x] Added changelog

TBlindaruk committed:
[9.x] Added changelog

BrandonSurowiec committed:
Update CHANGELOG.md (#43002)

The Str::mask PR for 9.x was already some versions ago. Removing the reference to the 8.x PR.

mateusjunges committed:
add tests for whereNot database rule (#43003)

Commit:
quote from Mustafa Kemal Atatürk (#43022)

wfeller committed:
[9.x] Fix missing return (#43023)

* Fix missing return

* add test

Co-authored-by: Tim MacDonald <hello@timacdonald.me>

browner12 committed:
allow Collection `random()` to accept a callable (#43028)

this allows the user to, rather than passing an int directly, pass a closure that can reference back to the original Collection in case they need to reference the size of the Collection

ralphjsmit committed:
Add App::forgetInstance() docblock (#43031)

StyleCIBot committed:
Apply fixes from StyleCI

BrandonSurowiec committed:
Disable Column Statistics for `php artisan schema:dump` on MariaDB (#43027)

Command fails on v10.x MariaDB.

lucasmichot committed:
Fix setUp and tearDown visibility. (#43052)

lucasmichot committed:
Simplify sone FQCN. (#43053)

ksassnowski committed:
Fix inconsistent content type when using ResponseSequence (#43051)

Commit:
[9.x] Bind a Vite Null Object to the Container instead of a Closure (#43040)

* Bind a Vite Null Object to the Container instead of a Closure

* Formatting

* Update src/Illuminate/Foundation/Testing/Concerns/InteractsWithContainer.php

Co-authored-by: Jess Archer <jess@jessarcher.com>

* Fix invoking of Vite test instance

* Add test for reactRefresh method on Vite test instance

Co-authored-by: Taylor Otwell <taylor@laravel.com>
Co-authored-by: Jess Archer <jess@jessarcher.com>

taylorotwell committed:
update wording

chu121su12 commented:
I'm on some kind of break. I'll retry/update this sometime soon.

azizramdan committed:
Add `query()` method to docblock DB facade (#43071)

rodrigopedra committed:
[8.x] Prevent double throwing chained exception on sync queue (#42950)

* prevent double throwing chained exception on sync queue

* decrease job count after processing

fragkp committed:
add int type to $key parameter for contains method (#43094)

danharrin committed:
[9.x] feature: `Str::inlineMarkdown()` (#43126)

* feature: Inline markdown helpers

* chore: fix styleci

* Update Str.php

Co-authored-by: Taylor Otwell <taylor@laravel.com>

Commit:
Avoid deprecation-warnings in Str::contains with PHP 8.1 (#43125)

* Avoid deprecation-warnings

not sure if this should be added in front, as it would be true, but the method is only for strings :shrug: 
        if ($haystack === $needles) {
            return true;
        }

* Update Str.php

Co-authored-by: Taylor Otwell <taylor@laravel.com>

erikn69 committed:
Added required directive (#43103)

Sergiobop committed:
[9.x] Add replicateQuietly to Model (#43141)

* Add replicateQuietly method

* Add test for replicateQuietly()

n8man committed:
Avoid matching multi-line imports (#43093)

Multi-line imports are broken:
```
use Illuminate\Foundation\Testing\{
    DatabaseTransactions,
    WithFaker,
};
```
Results in output:
```
    DatabaseTransactions,
    WithFaker,
use Illuminate\Foundation\Testing\{
};
```

timacdonald committed:
[9.x] Allow authorization responses to specify HTTP status codes (#43097)

* allow authorize response to specify http status

* code style

* add 404 helpers

erikn69 committed:
Early return when message format is the default (#43149)

driesvints committed:
Revert "Avoid deprecation-warnings in Str::contains with PHP 8.1 (#43125)" (#43164)

This reverts commit adcbc30f2cfb7fb17d74b23c8ef0d62b1fcf73f5.

valorin committed:
Add ignore param to ValidateSignature middleware (#43160)

* Add ignore param to ValidateSignature middleware

* Add test for ValidateSignature $ignore param

StyleCIBot committed:
Apply fixes from StyleCI

taylorotwell committed:
version

taylorotwell committed:
fix conflicts

chu121su12 committed:
[9.x] Add support for `Artisan::command('cli', [ClassName::class, '__invoke'])`

chu121su12 committed:
Merge remote-tracking branch 'github/feat/callable-command' into feat/callable-command

chu121su12 commented:
Tracked the error on prefer-lowest test fail due to changes on serializer
https://github.com/laravel/serializable-closure/compare/v1.0.5...v1.1.0

chu121su12 commented on src/Illuminate/Foundation/Console/Kernel.php:
```diff
@@ -185,12 +185,14 @@ protected function commands()
+    public function command($signature, $callback)
     {
-        $command = new ClosureCommand($signature, $callback);
+        $command = $callback instanceof Closure
+            ? new ClosureCommand($signature, $callback)
+            : new CallableCommand($signature, $callback);
```
I tried this logic and the code here

https://github.com/laravel/framework/blob/1f77f3d586debda17be6a886b172e7336d65bb60/src/Illuminate/Foundation/Console/CallableCommand.php#L34-L77

is quite difficult to be spliced around this section.
